File: bindings/csharp/conformance/Pm/PmSteps.cs
using System.Collections.Generic;
using Angzarr;
using Angzarr.Router;
using NUnit.Framework;
using Reqnroll;
using TC = Test.Counter;

namespace Angzarr.Router.Conformance.Pm;

/// <summary>Step definitions for process_manager.feature — the
/// OrderProcessManager stateful trigger-side dispatch.</summary>
[Binding]
[Scope(Feature = "Order process-manager dispatch")]
public sealed class PmSteps
{
    private Router _router = null!;
    private ProcessManagerHandleResponse? _resp;
    private CodedError? _err;

    [BeforeScenario]
    public void Before()
    {
        _router = new Router();
        TC.counter_angzarr.RegisterOrderProcessManager(_router, new PmFixture());
        _resp = null;
        _err = null;
    }

    [AfterScenario]
    public void After() => _router?.Dispose();

    private void Dispatch(ProcessManagerHandleRequest req)
    {
        try
        {
            _resp = _router.DispatchProcessManager(req);
            _err = null;
        }
        catch (CodedError e)
        {
            _err = e;
            _resp = null;
        }
    }

    [Given("an order process-manager")]
    public void AnOrderProcessManager()
    {
        // The fixture is registered in Before.
    }

    [When(
        "an Increased trigger in domain {string} is dispatched with destination inventory sequence {int}"
    )]
    public void IncreasedWithDestination(string domain, int seq) =>
        Dispatch(
            Builders.PmTrigger(
                domain,
                new[] { "test.counter.Increased" },
                null,
                new Dictionary<string, uint> { ["inventory"] = (uint)seq }
            )
        );

    [When("an Increased trigger in domain {string} is dispatched")]
    public void IncreasedInDomain(string domain) =>
        Dispatch(Builders.PmTrigger(domain, new[] { "test.counter.Increased" }, null, null));

    [When("a trigger whose newest page is an undeclared event is dispatched")]
    public void NewestUndeclared() =>
        Dispatch(
            Builders.PmTrigger(
                "counter",
                new[] { "test.counter.Increased", "test.counter.Unwatched" },
                null,
                null
            )
        );

    [When("an Increased trigger is dispatched over a prior state of {int} events")]
    public void IncreasedOverState(int n) =>
        Dispatch(
            Builders.PmTrigger(
                "counter",
                new[] { "test.counter.Increased" },
                Builders.PmStateOf(n),
                null
            )
        );

    [When("a request with no trigger is dispatched")]
    public void NoTrigger() => Dispatch(Builders.PmNoTrigger());

    [When("a trigger with no pages is dispatched")]
    public void EmptyTrigger() => Dispatch(Builders.PmEmptyTrigger());

    [When("a rejection of Reserve is dispatched")]
    public void RejectionReserve() => Dispatch(Builders.PmRejection("test.counter.Reserve"));

    [Then("the process-manager emits one command to {string}")]
    public void EmitsOneCommand(string target)
    {
        Assert.That(_err, Is.Null, "dispatch unexpectedly failed");
        Assert.That(_resp!.Commands.Count, Is.EqualTo(1), "emitted commands");
        Assert.That(_resp.Commands[0].Cover.Domain, Is.EqualTo(target), "command target");
    }

    [Then("the command carries destination sequence {int}")]
    public void CommandCarriesSequence(int seq) =>
        Assert.That((int)_resp!.Commands[0].Pages[0].Header.Sequence, Is.EqualTo(seq));

    [Then("the process-manager emits no commands")]
    public void EmitsNoCommands()
    {
        Assert.That(_err, Is.Null, "dispatch unexpectedly failed");
        Assert.That(_resp!.Commands.Count, Is.EqualTo(0), "expected no commands");
    }

    [Then("the process-manager rebuilt {int} prior state events")]
    public void RebuiltN(int n)
    {
        Assert.That(_err, Is.Null, "dispatch unexpectedly failed");
        Assert.That(_resp!.Facts.Count, Is.EqualTo(n), "rebuilt prior state events");
    }

    [Then("the dispatch fails with {word}")]
    public void DispatchFailsWith(string code)
    {
        Assert.That(_err, Is.Not.Null, "expected coded error " + code);
        Assert.That(_err!.Code, Is.EqualTo(code));
    }

    [Then("the process-manager emits one process event")]
    public void EmitsOneProcessEvent()
    {
        Assert.That(_err, Is.Null, "dispatch unexpectedly failed");
        Assert.That(_resp!.ProcessEvents.Count, Is.EqualTo(1), "process events");
    }

    [Then("the process-manager escalates")]
    public void Escalates()
    {
        Assert.That(_err, Is.Null, "dispatch unexpectedly failed");
        Assert.That(_resp!.Notification, Is.Not.Null, "expected an escalation");
    }
}

File: bindings/java/src/test/java/io/angzarr/router/conformance/pm/PMSteps.java
package io.angzarr.router.conformance.pm;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.angzarr.ProcessManagerHandleRequest;
import io.angzarr.ProcessManagerHandleResponse;
import io.angzarr.router.CodedError;
import io.angzarr.router.Router;
import io.angzarr.router.conformance.Builders;
import io.cucumber.java.After;
import io.cucumber.java.Before;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import java.util.List;
import java.util.Map;
import test.counter.counter_angzarr;

/** Step definitions for process_manager.feature — the OrderProcessManager
 * stateful trigger-side dispatch. */
public class PMSteps {

  private Router router;
  private ProcessManagerHandleResponse resp;
  private CodedError err;

  @Before
  public void before() {
    router = new Router();
    counter_angzarr.registerOrderProcessManager(router, new PMFixture());
    resp = null;
    err = null;
  }

  @After
  public void after() {
    if (router != null) {
      router.close();
    }
  }

  private void dispatch(ProcessManagerHandleRequest req) {
    try {
      resp = router.dispatchProcessManager(req);
      err = null;
    } catch (CodedError e) {
      err = e;
      resp = null;
    }
  }

  @Given("an order process-manager")
  public void anOrderProcessManager() {
    // The fixture is registered in @Before.
  }

  @When("an Increased trigger in domain {string} is dispatched with destination inventory sequence {int}")
  public void increasedWithDestination(String domain, int seq) {
    dispatch(
        Builders.pmTrigger(
            domain, List.of("test.counter.Increased"), null, Map.of("inventory", seq)));
  }

  @When("an Increased trigger in domain {string} is dispatched")
  public void increasedInDomain(String domain) {
    dispatch(Builders.pmTrigger(domain, List.of("test.counter.Increased"), null, null));
  }

  @When("a trigger whose newest page is an undeclared event is dispatched")
  public void newestUndeclared() {
    dispatch(
        Builders.pmTrigger(
            "counter", List.of("test.counter.Increased", "test.counter.Unwatched"), null, null));
  }

  @When("an Increased trigger is dispatched over a prior state of {int} events")
  public void increasedOverState(int n) {
    dispatch(
        Builders.pmTrigger("counter", List.of("test.counter.Increased"), Builders.pmStateOf(n), null));
  }

  @When("a request with no trigger is dispatched")
  public void noTrigger() {
    dispatch(Builders.pmNoTrigger());
  }

  @When("a trigger with no pages is dispatched")
  public void emptyTrigger() {
    dispatch(Builders.pmEmptyTrigger());
  }

  @When("a rejection of Reserve is dispatched")
  public void rejectionReserve() {
    dispatch(Builders.pmRejection("test.counter.Reserve"));
  }

  @Then("the process-manager emits one command to {string}")
  public void emitsOneCommand(String target) {
    assertNull(err, "dispatch failed");
    assertEquals(1, resp.getCommandsList().size(), "emitted commands");
    assertEquals(target, resp.getCommandsList().get(0).getCover().getDomain(), "command target");
  }

  @Then("the command carries destination sequence {int}")
  public void commandCarriesSequence(int seq) {
    assertEquals(seq, resp.getCommandsList().get(0).getPages(0).getHeader().getSequence());
  }

  @Then("the process-manager emits no commands")
  public void emitsNoCommands() {
    assertNull(err, "dispatch failed");
    assertEquals(0, resp.getCommandsList().size(), "expected no commands");
  }

  @Then("the process-manager rebuilt {int} prior state events")
  public void rebuiltN(int n) {
    assertNull(err, "dispatch failed");
    assertEquals(n, resp.getFactsList().size(), "rebuilt prior state events");
  }

  @Then("the dispatch fails with {word}")
  public void dispatchFailsWith(String code) {
    assertNotNull(err, "expected coded error " + code);
    assertEquals(code, err.code);
  }

  @Then("the process-manager emits one process event")
  public void emitsOneProcessEvent() {
    assertNull(err, "dispatch failed");
    assertEquals(1, resp.getProcessEventsList().size(), "process events");
  }

  @Then("the process-manager escalates")
  public void escalates() {
    assertNull(err, "dispatch failed");
    assertTrue(resp.hasNotification(), "expected an escalation");
  }
}
